src/renderer/components/+extensions/extensions.tsx

```diff
@@ -32,27 +54,238 @@ export class Extensions extends React.Component {
+  getExtensionDestFolder(name: string) {
+    return path.join(this.extensionsPath, sanitizeExtensionName(name));
+  }
+
+  installFromSelectFileDialog = async () => {
+    const { dialog, BrowserWindow, app } = remote;
+    const { canceled, filePaths } = await dialog.showOpenDialog(BrowserWindow.getFocusedWindow(), {
+      defaultPath: app.getPath("downloads"),
+      properties: ["openFile", "multiSelections"],
+      message: _i18n._(t`Select extensions to install (formats: ${this.supportedFormats.join(", ")}), `),
+      buttonLabel: _i18n._(t`Use configuration`),
+      filters: [
+        { name: "tarball", extensions: this.supportedFormats }
+      ]
+    });
+    if (!canceled && filePaths.length) {
+      this.requestInstall(
+        filePaths.map(filePath => ({
+          fileName: path.basename(filePath),
+          filePath: filePath,
+        }))
+      );
+    }
+  };
+
+  installExtensions = () => {
+    if (this.downloadUrl) {
+      this.installFromNpmOrUrl(this.downloadUrl);
+      this.downloadUrl = "";
+    } else {
+      this.installFromSelectFileDialog();
+    }
+  };
+
+  installFromNpmOrUrl = async (url = this.downloadUrl) => {
+    if (!InputValidators.isUrl.validate(url)) {
+      const npmPackageName = url;
+      url = extensionManager.getNpmPackageTarballUrl(npmPackageName);
+      if (!url) {
+        Notifications.error(`Error: npm package "${npmPackageName}" not found!`);
+        return;
+      }
+    }
+    try {
+      const { promise: filePromise } = downloadFile({ url });
+      this.requestInstall([{
+        fileName: path.basename(url),
+        data: await filePromise,
+      }]);
+    } catch (err) {
+      Notifications.error(
+        <div className="flex column gaps">
+          <p>Installation from URL has failed: <b>{String(err)}</b></p>
+          <p>URL: <em>{url}</em></p>
+        </div>
+      );
+    }
+  };
+
+  installOnDrop = (files: File[]) => {
+    logger.info('Install from D&D');
+    return this.requestInstall(
+      files.map(file => ({
+        fileName: path.basename(file.path),
+        filePath: file.path,
+      }))
+    );
+  };
+
+  async requestInstall(installRequests: InstallRequest[]) {
+    const pendingFiles: Promise<any>[] = [];
+
+    // read extensions with provided system path if any
+    installRequests.forEach(ext => {
+      if (ext.data) return;
+      const promise = fse.readFile(ext.filePath)
+        .then(data => ext.data = data)
+        .catch(err => {
+          Notifications.error(`Error while reading "${ext.filePath}": ${String(err)}`);
+        });
+      pendingFiles.push(promise);
+    });
+    await Promise.all(pendingFiles);
+    installRequests = installRequests.filter(item => item.data); // remove items with reading errors
+
+    // prepare temp folder
+    const tempFolder = path.join(os.tmpdir(), "lens-extensions");
+    await fse.ensureDir(tempFolder);
+
+    // copy files to temp, get extension info from package.json and do basic validation
```

Comment:
This function is quite long and does many things, I'd split it into 2 or 3 functions.

Reply:
Yes